## E2E is red at `a480125a` — and "it is the known flake" is a hypothesis you have to test, not assume

Khaliq flagged the failing check. Per-workflow at head `a480125a`: **9 of 10 green, `E2E Tests` failed.** The same job passed at the previous head `a8e5f157` on this branch 24 minutes earlier.

The failure signature:

```
Graceful shutdown timed out after 10000ms. Use --force to kill.
[ERROR] Broker still reported as running after down command
Process completed with exit code 1
```

**Why the obvious answer is not good enough here.** That signature matches the known E2E flake — relay#1550 records `E2E Tests` failing on 4 of the last 15 pushes to main, and the relay#1543 lane hit this exact "graceful shutdown timed out" string this morning, re-ran in isolation, and it passed. There is also an unowned E2E shutdown fix sitting at `160d5a2c2`.

**But this PR is the worst possible one to wave that away on.** You added a reconciliation loop that runs in the broker, with a bounded retry and a 60-second backoff. "Broker still reported as running after down command" is *exactly* what a background task that is never joined or aborted on shutdown looks like. A flaky test and a real shutdown regression produce an identical line here.

So do both, and report both:

1. **Re-run `E2E Tests` in isolation** at the same head. If it goes green, that is evidence for flake — but only evidence, not proof, because a race that fires 25% of the time and a race your change made slightly more likely are indistinguishable from one green re-run.
2. **Affirmatively check the shutdown path.** Is the reconciliation task joined, aborted, or cancelled when the broker shuts down? Name the file and line that does it. If nothing does, you have found a real regression and the flake was a red herring — and that is the more valuable outcome of the two.

Point 2 is the one that matters. **A green re-run answers "did this test pass"; it does not answer "does my task hold the broker open."** Only reading the shutdown path answers that, and it takes minutes.

The prior head passing with the reconciler already present is mild evidence against the regression theory, and I am recording it as mild rather than conclusive — a timing-dependent hold would not fail every run.

State otherwise: `mergeable: MERGEABLE`, everything else green. **Live threads still unanswered** — the generation-vs-PID P2 at `fleet.rs:1743`, `worker.rs:249`, and `fleet.rs:1806` where you claim `73136fad2` addressed it but left it unresolved. Those plus this E2E answer are the whole remaining gate.

--- a/crates/broker/src/runtime/fleet.rs
+++ b/crates/broker/src/runtime/fleet.rs
@@ -31,7 +31,7 @@ const FLEET_INVENTORY_RECONCILE_FAILURE_BACKOFF: Duration = Duration::from_secs(
 
 #[derive(Debug, Clone, Copy, PartialEq, Eq)]
 pub(super) struct FleetInventoryRetry {
-    process_id: u32,
+    generation: Uuid,
     retry_after: Instant,
 }
 // Relaycast currently limits a node to 32 terminal sessions. Keep that many
@@ -1717,13 +1717,13 @@ pub(super) async fn record_fleet_inventory_agent(
 fn schedule_fleet_inventory_retry(
     retry_after: &mut HashMap<WorkerName, FleetInventoryRetry>,
     name: WorkerName,
-    process_id: u32,
+    generation: Uuid,
     now: Instant,
 ) {
     retry_after.insert(
         name,
         FleetInventoryRetry {
-            process_id,
+            generation,
             retry_after: now + FLEET_INVENTORY_RECONCILE_FAILURE_BACKOFF,
         },
     );
@@ -1738,15 +1738,15 @@ pub(super) async fn reconcile_fleet_inventory_with_live_workers(
     live_workers: Vec<LiveFleetInventoryCandidate>,
     now: Instant,
 ) -> usize {
-    let live_worker_processes: HashMap<_, _> = live_workers
+    let live_worker_generations: HashMap<_, _> = live_workers
         .iter()
-        .map(|worker| (worker.name.clone(), worker.process_id))
+        .map(|worker| (worker.name.clone(), worker.generation))
         .collect();
     // A worker that exits or has since been restored needs no retained retry
-    // state. A restarted same-name worker has a different PID and must not
+    // state. A restarted same-name worker has a different generation and must not
     // inherit the old process's retry deadline.
     retry_after.retain(|name, retry| {
-        live_worker_processes.get(name) == Some(&retry.process_id)
+        live_worker_generations.get(name) == Some(&retry.generation)
             && !fleet_inventory.contains_key(name)
     });
     let missing_workers: Vec<_> = live_workers
@@ -1773,7 +1773,7 @@ pub(super) async fn reconcile_fleet_inventory_with_live_workers(
     for LiveFleetInventoryCandidate {
         name,
         session_ref,
-        process_id,
+        generation,
     } in missing_workers
     {
         let agent = match timeout(
@@ -1790,7 +1790,7 @@ pub(super) async fn reconcile_fleet_inventory_with_live_workers(
                     retry_after_secs = FLEET_INVENTORY_RECONCILE_FAILURE_BACKOFF.as_secs(),
                     "could not resolve live worker for fleet inventory reconciliation; will retry later"
                 );
-                schedule_fleet_inventory_retry(retry_after, name, process_id, now);
+                schedule_fleet_inventory_retry(retry_after, name, generation, now);
                 continue;
             }
             Err(_) => {
@@ -1800,7 +1800,7 @@ pub(super) async fn reconcile_fleet_inventory_with_live_workers(
                     retry_after_secs = FLEET_INVENTORY_RECONCILE_FAILURE_BACKOFF.as_secs(),
                     "timed out resolving live worker for fleet inventory reconciliation; will retry later"
                 );
-                schedule_fleet_inventory_retry(retry_after, name, process_id, now);
+                schedule_fleet_inventory_retry(retry_after, name, generation, now);
                 continue;
             }
         };
@@ -1811,7 +1811,7 @@ pub(super) async fn reconcile_fleet_inventory_with_live_workers(
                 retry_after_secs = FLEET_INVENTORY_RECONCILE_FAILURE_BACKOFF.as_secs(),
                 "refusing to reconcile fleet inventory with a mismatched Relaycast identity; will retry later"
             );
-            schedule_fleet_inventory_retry(retry_after, name, process_id, now);
+            schedule_fleet_inventory_retry(retry_after, name, generation, now);
             continue;
         }
 
@@ -1824,7 +1824,7 @@ pub(super) async fn reconcile_fleet_inventory_with_live_workers(
                     retry_after_secs = FLEET_INVENTORY_RECONCILE_FAILURE_BACKOFF.as_secs(),
                     "refusing to replace a live worker's authoritative fleet identity; will retry later"
                 );
-                schedule_fleet_inventory_retry(retry_after, name, process_id, now);
+                schedule_fleet_inventory_retry(retry_after, name, generation, now);
                 continue;
             }
         } else {
@@ -2277,12 +2277,12 @@ mod tests {
     fn live_fleet_worker(
         name: &str,
         session_ref: Option<&str>,
-        process_id: u32,
+        generation: u128,
     ) -> LiveFleetInventoryCandidate {
         LiveFleetInventoryCandidate {
             name: WorkerName::from(name),
             session_ref: session_ref.map(ToOwned::to_owned),
-            process_id,
+            generation: Uuid::from_u128(generation),
         }
     }
 
@@ -3409,7 +3409,7 @@ mod tests {
         assert_eq!(
             retry_after.get(&WorkerName::from("live-worker")),
             Some(&FleetInventoryRetry {
-                process_id: 103,
+                generation: Uuid::from_u128(103),
                 retry_after: now + FLEET_INVENTORY_RECONCILE_FAILURE_BACKOFF,
             })
         );
@@ -3460,7 +3460,7 @@ mod tests {
             worker_names
                 .iter()
                 .enumerate()
-                .map(|(index, name)| live_fleet_worker(name, None, index as u32 + 1))
+                .map(|(index, name)| live_fleet_worker(name, None, index as u128 + 1))
                 .collect()
         };
         let now = Instant::now();
@@ -3519,7 +3519,7 @@ mod tests {
         let mut inventory = HashMap::new();
         let mut delivery_book = FleetDeliveryBook::default();
         let mut retry_after = HashMap::new();
-        let live_workers = |process_id| vec![live_fleet_worker("missing-worker", None, process_id)];
+        let live_workers = |generation| vec![live_fleet_worker("missing-worker", None, generation)];
         let now = Instant::now();
 
         assert_eq!(
@@ -3539,7 +3539,7 @@ mod tests {
         assert_eq!(
             retry_after.get(&WorkerName::from("missing-worker")),
             Some(&FleetInventoryRetry {
-                process_id: 104,
+                generation: Uuid::from_u128(104),
                 retry_after: now + FLEET_INVENTORY_RECONCILE_FAILURE_BACKOFF,
             })
         );
@@ -3580,7 +3580,7 @@ mod tests {
         assert_eq!(
             retry_after.get(&WorkerName::from("missing-worker")),
             Some(&FleetInventoryRetry {
-                process_id: 105,
+                generation: Uuid::from_u128(105),
                 retry_after: now
                     + Duration::from_secs(2)
                     + FLEET_INVENTORY_RECONCILE_FAILURE_BACKOFF,
@@ -3649,7 +3649,7 @@ mod tests {
         assert_eq!(
             retry_after.get(&WorkerName::from("slow-worker")),
             Some(&FleetInventoryRetry {
-                process_id: 106,
+                generation: Uuid::from_u128(106),
                 retry_after: now + FLEET_INVENTORY_RECONCILE_FAILURE_BACKOFF,
             })
         );

--- a/crates/broker/src/worker.rs
+++ b/crates/broker/src/worker.rs
@@ -242,9 +242,9 @@ pub(crate) enum WorkerEvent {
 pub(crate) struct LiveFleetInventoryCandidate {
     pub(crate) name: WorkerName,
     pub(crate) session_ref: Option<String>,
-    /// The wrapper PID distinguishes a restarted same-name worker from the
+    /// The process generation distinguishes a restarted same-name worker from the
     /// process whose failed Relaycast lookup is currently being backed off.
-    pub(crate) process_id: u32,
+    pub(crate) generation: Uuid,
 }
 
 pub(crate) struct WorkerRegistry {
@@ -522,7 +522,6 @@ impl WorkerRegistry {
                 if handle.parent.is_none() || !self.is_worker_live(name) {
                     return None;
                 }
-                let process_id = handle.child.id()?;
                 let session_ref = handle.spec.session_id.clone().or_else(|| {
                     handle
                         .spec
@@ -534,7 +533,7 @@ impl WorkerRegistry {
                 Some(LiveFleetInventoryCandidate {
                     name: name.clone(),
                     session_ref,
-                    process_id,
+                    generation: handle.generation,
                 })
             })
             .collect()

